### PR TITLE
refactor: support run zipper from a file

### PR DESCRIPTION
**Pull request by woorui**
# Description

1. Support Zipper launch from a config file.

2. Remove the `-a -m` flag from `yomo serve`, as all configurations are now covered by the configuration file.

3. If you want to develop a Zipper, use `yomo.NewZipper()`. If you want to run a Zipper from a configuration, use `yomo.RunZipper()`.

### CODE REVIEW

File: example/4-cascading-zipper/zipper-1/zipper_1.go [hunk 1]
```diff
@@ -1,35 +1,41 @@
...
+	conf, err := config.ParseWorkflowConfig("zipper_1_wf.yaml")
 	if err != nil {
-		panic(err)
+		log.Fatalln(err)
+	}
+
+	zipper, err := yomo.NewZipper(
+		conf.Name,
+		conf.Functions,
+		yomo.WithDownstreamOption(yomo.WithAuth("token", "z1")),
```
Comment (venjiang) [inline example/4-cascading-zipper/zipper-1/zipper_1.go:22]: 'yomo. WithDownstreamOption' confused me.

Comment (fanweixiao) [reply example/4-cascading-zipper/zipper-1/zipper_1.go:22]: should remove this from `NewZipper()`, we have `AddDownstreamZipper()` to do this.

File: core/server.go [hunk 2]
```diff
@@ -81,6 +77,15 @@ func (s *Server) ListenAndServe(ctx context.Context, addr string) error {
 		return err
 	}
 
+	s.logger = s.logger.With("zipper_addr", addr)
+
+	// connect to all downstreams.
+	for addr, client := range s.downstreams {
+		if err := client.Connect(ctx, addr); err != nil {
```
[the remark this answers is not in the document]
Comment (woorui) [reply core/server.go:84]: https://github.com/yomorun/yomo/blob/972cd3fe607d94c996379bf20e8cd9c51821f7ee/core/client.go#L76

The `Connect` function only returns an error if the first connection attempt fails. Reconnection is then automatically attempted in the background.

File: example/4-cascading-zipper/zipper-1/zipper_1.go (hunk 1 above)
Comment (woorui) [reply example/4-cascading-zipper/zipper-1/zipper_1.go:22]: The Zipper has two types: Downstream and Upstream. Use `DownstreamOption` for the Downstream Zipper and `UpstreamOption` for the Upstream Zipper.

File: zipper_windows.go [hunk 3]
```diff
@@ -13,17 +13,16 @@ import (
...
-				os.Exit(0)
-			}
+func waitSignalForShotdownServer(server *core.Server) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, syscall.SIGTERM, syscall.SIGINT)
+	ylog.Info("Listening SIGTERM/SIGINT...")
+	for p1 := range c {
+		ylog.Debug("Received signal", "signal", p1)
+		if p1 == syscall.SIGTERM || p1 == syscall.SIGINT {
+			// server.Close()
```
[the remark this answers is not in the document]
Comment (venjiang) [reply zipper_windows.go:23]: I think this is a mistake and this line of code should be added. 😄

File: example/3-multi-sfn/zipper/main.go
Comment (wujunzhuo) [inline example/3-multi-sfn/zipper/main.go:1]: I suggest delete this file, and start zipper server from cli directly.

File: pkg/config/config.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+// Downstream describes a cascading zipper config.
+type Downstream struct {
+	// Name is the name of downstream zipper.
+	Name string `yaml:"name"`
+	// Host is the host of downstream zipper.
+	Host string `yaml:"host"`
+	// Port is the port of downstream zipper.
+	Port int `yaml:"port"`
+	// Credential is the credential that specifies how downstream will authenticate the current Zipper.
```
Comment (fanweixiao) [inline pkg/config/config.go:44]:
```suggestion
	// Credential is the credential when connect to downstream zipper.
```

File: pkg/config/config.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	// It is in the format of 'authType:authPayload', separated by a colon.
+	// If Credential is empty, it represents that downstream will not authenticate the current Zipper.
+	Credential string `yaml:"credential"`
+}
+
+// ErrConfigExt represents the extension of config file is incorrect.
+var ErrConfigExt = errors.New(`yomo: the extension of config is incorrect, it should ".yaml|.yml"`)
```
Comment (fanweixiao) [inline pkg/config/config.go:51]:
```suggestion
var ErrConfigExt = errors.New(`yomo: the extension of config file is incorrect, it should be ".yaml|.yml"`)
```

File: pkg/config/config.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	if conf.Port == 0 {
+		return errors.New("config: the port is required")
+	}
+	if len(conf.Functions) == 0 {
+		return errors.New("config: the functions cannot be an empty")
+	}
+
+	for _, f := range conf.Functions {
+		if f.Name == "" {
+			return errors.New("config: the functions must have name value")
```
Comment (fanweixiao) [inline pkg/config/config.go:92]:
```suggestion
			return errors.New("config: the functions must have the name field")
```

**Review by fanweixiao — CHANGES_REQUESTED** [review]

**Review by fanweixiao — APPROVED** [review]
👍🏼

**Review by fanweixiao — APPROVED** [review]

File: zipper_notwindows.go [hunk 7]
```diff
@@ -9,32 +9,31 @@ import (
...
-				z.Close()
-				os.Exit(0)
-			} else if p1 == syscall.SIGUSR2 {
-				var m runtime.MemStats
-				runtime.ReadMemStats(&m)
-				ylog.Debug("runtime stats", "gc_nums", m.NumGC)
-			} else if p1 == syscall.SIGUSR1 {
-				ylog.Debug("zipper stats", "zipper_stats", z.Stats())
-			}
+func waitSignalForShotdownServer(server *core.Server) {
```
Comment (wujunzhuo) [inline zipper_notwindows.go:20]: shutdown

File: test/config.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+### general ###
+name: zipper-chn
```
Comment (woorui) [inline test/config.yaml:2]: I standardized the naming convention for local name and downstreams name. @wujunzhuo

**Review by venjiang — APPROVED** [review]
👏

**Pull request merged**